tket/tests/test_LexiRoute.cpp

```diff
@@ -1739,4 +1740,228 @@ SCENARIO(
   mapping_frontier.circuit_.get_commands();
   REQUIRE(true);
 }
+
+SCENARIO(
+    "Test case fails if the linear boundary in merge_ancilla replaces the "
+    "ancilla entry before erasing the merge entry.") {
+  SquareGrid architecture(5, 5);
+  std::vector<Node> nodes = architecture.get_all_nodes_vec();
+  Circuit circuit(10);
+  circuit.add_op<unsigned>(OpType::CX, {0, 1});
+  circuit.add_op<unsigned>(OpType::CX, {0, 3});
+  circuit.add_op<unsigned>(OpType::CX, {1, 2});
+  circuit.add_barrier({0, 1, 2, 3, 4, 5, 6, 7, 8, 9});
+  circuit.add_op<unsigned>(OpType::CX, {6, 7});
+  std::map<Qubit, Node> p_map = {
+      // mapping for qbs with 2qb gates
+      {Qubit(0), nodes[0]},  {Qubit(1), nodes[4]},  {Qubit(2), nodes[20]},
+      {Qubit(3), nodes[24]}, {Qubit(4), nodes[11]}, {Qubit(5), nodes[17]},
+  };
+  Placement::place_with_map(circuit, p_map);
+  CompilationUnit cu(circuit);
+  PassPtr r_p = gen_routing_pass(
+      architecture, {std::make_shared<LexiLabellingMethod>(),
+                     std::make_shared<LexiRouteRoutingMethod>()});
+  REQUIRE(r_p->apply(cu));
+}
+
+SCENARIO(
+    "Test relabelling a Circuit UnitID that is an Architecture Node but "
+    "reassignable to an Ancilla Node.") {
+  GIVEN("Line Architecture, one reassignment.") {
```

Comment:
Can you add some comments explaining why some UnitID have been reassigned to an ancilla? My next comment is related to this.

Reply:
Unfortunately the condition in which this occurs is rather niche. I've added a small explanation to the test:
• If a non-labelled Qubit in a Circuit being mapped has no Quantum gates with physical constraints (i.e. mostly multi-qubit gates) then before mapping we assign it some "bad" Architecture Node (typically something on the edge of the coupling graph with low out-degree)
• Any non-labelled Qubit in a Circuit with Quantum gates with physical constraints are left unlabelled
• During mapping, if a multi-qubit gate with a non-labelled Qubit is encountered we need to allocate it to some _best_ Architecture Node
• In some cases, this _best_ Node may end up being a "bad" Node we've used to assign an "unimportant" (without connectivity graph related physical constraints) Qubit too. If this is the case we relabel the unlabelled Qubit to this _best_ Node and find a new Node to assign the "unimportant" Qubit to
• Ideally we just find a spare Architecture Node that hasn't previously been assigned to and reassign the "unimportant" Qubit to it. However in some cases there can be no spare Architecture Node as they have been used as ancilla Node for SWAP/BRIDGE gates
• This is the case that was not considered before - the assertion is basically saying "there should always be a spare Architecture Node" but this is not true **(this isn't in the comments)**
• In this case we take an Ancilla Node and wire its output to the input of the "unimportant" Qubit Path, essentially reassigning it.

Reply:
Thanks! this is much appreciated.